packages/protocol/contracts/ACE/noteRegistry/epochs/201907/adjustable/Behaviour.sol

```diff
@@ -0,0 +1,106 @@
+
+    /**
+    * @dev Burn AZTEC notes
+    * TODO
+    */
+    function burn(bytes calldata _proofOutputs) external onlyOwner {
+        require(registry.canAdjustSupply == true, "this asset is not burnable");
+        // Dealing with notes representing totals
+        (bytes memory oldTotal, // input notes
+        bytes memory newTotal, // output notes
+        ,
+        ) = _proofOutputs.get(0).extractProofOutput();
+
+        // Dealing with burned notes
+        (,
+        bytes memory burnedNotes,
+        ,) = _proofOutputs.get(1).extractProofOutput();
+
+        (, bytes32 oldTotalNoteHash, ) = oldTotal.get(0).extractNote();
+        require(oldTotalNoteHash == registry.confidentialTotalBurned, "provided total burned note does not match");
+        (, bytes32 newTotalNoteHash, ) = newTotal.get(0).extractNote();
+        setConfidentialTotalBurned(newTotalNoteHash);
+
+        // Although they are outputNotes, they are due to be destroyed - need removing from the note registry
+        updateInputNotes(burnedNotes);
+    }
+
+        /**
+    * @dev Mint AZTEC notes
+    * TODO
+    */
+    function mint(bytes calldata _proofOutputs) external onlyOwner {
+        require(registry.canAdjustSupply == true, "this asset is not mintable");
+        // Dealing with notes representing totals
+        (bytes memory oldTotal, // input notes
+        bytes memory newTotal, // output notes
+        ,
+        ) = _proofOutputs.get(0).extractProofOutput();
+
+        // Dealing with burned notes
+        (,
+        bytes memory mintedNotes,
+        ,) = _proofOutputs.get(1).extractProofOutput();
+
+        (, bytes32 oldTotalNoteHash, ) = oldTotal.get(0).extractNote();
+        require(oldTotalNoteHash == registry.confidentialTotalMinted, "provided total burned note does not match");
+        (, bytes32 newTotalNoteHash, ) = newTotal.get(0).extractNote();
+        setConfidentialTotalMinted(newTotalNoteHash);
+
+        updateOutputNotes(mintedNotes);
+    }
+
+    /**
+        * @dev Update the state of the note registry according to transfer instructions issued by a 
+        * zero-knowledge proof
+        *
+        * @param _proof - unique identifier for a proof
+        * @param _proofOutput - transfer instructions issued by a zero-knowledge proof
+    */
+    function updateNoteRegistry(
+        uint24 _proof,
+        bytes memory _proofOutput
+    ) public onlyOwner {
+        require(registry.active == true, "note registry does not exist for the given address");
+
+        (bytes memory inputNotes,
+        bytes memory outputNotes,
+        ,
+        int256 publicValue) = _proofOutput.extractProofOutput();
+
+        updateInputNotes(inputNotes);
+        updateOutputNotes(outputNotes);
+
+        // If publicValue != 0, enact a token transfer
+        // (publicValue < 0) => transfer from publicOwner to ACE
+        // (publicValue > 0) => transfer from ACE to publicOwner
+        if (publicValue != 0) {
```

Comment:
Ditto

Reply:
On input arg, see above comment.

This is required, since you need to explicitly fail if a valid proof has been sent with a kPublic !== 0, otherwise you will end up modifying noteRegistry state and potentially creating value out of thin air.

Reply:
Yeah you're right